packages/salesforcedx-vscode-core/src/commands/forceSourceDeploy.ts

```diff
@@ -5,37 +5,129 @@
+  type: string;
+}
+
+interface ForceSourceDeployErrorResult {
+  message: string;
+  name: string;
+  result: Result[];
+  stack: string;
+  status: number;
+  warnings: any[];
+}
+
+// move this elsewhere?
+vscode.workspace.onDidChangeTextDocument((e => {
+  if (ForceSourceDeployExecutor._errorCollection.has(e.document.uri)) {
+    ForceSourceDeployExecutor._errorCollection.delete(e.document.uri);
+  }
+}));
+
+export class ForceSourceDeployExecutor extends SfdxCommandletExecutor<SelectedPath> {
+
+  public static _errorCollection = vscode.languages.createDiagnosticCollection('deploy-errors');
+
   public build(data: SelectedPath): Command {
     const commandBuilder = new SfdxCommandBuilder()
       .withDescription(nls.localize('force_source_deploy_text'))
       .withArg('force:source:deploy')
-      .withLogName('force_source_deploy');
+      .withLogName('force_source_deploy')
+      .withJson();
     if (data.type === FileType.Manifest) {
       commandBuilder.withFlag('--manifest', data.filePath);
     } else {
       commandBuilder.withFlag('--sourcepath', data.filePath);
     }
     return commandBuilder.build();
   }
+
+  public execute(response: ContinueResponse<SelectedPath>): void {
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+
+    const execution = new CliCommandExecutor(this.build(response.data), {
+      cwd: vscode.workspace.rootPath
+    }).execute(cancellationToken);
+
+    let stdErr = '';
+    execution.stderrSubject.subscribe(realData => {
+      stdErr += realData.toString();
+    });
+
+    execution.processExitSubject.subscribe(async exitCode => {
+      if (exitCode !== 0) {
+        try {
+          const compileErrs = this.getDeployResultData(stdErr);
+          const fileErrors = this.groupErrorsByPath(compileErrs);
+          Object.keys(fileErrors).forEach(filePath => {
+            const uri = vscode.Uri.file(path.join(vscode.workspace.rootPath || '', filePath));
+            const fileDiagnostics: vscode.Diagnostic[] = [];
+            fileErrors[filePath].forEach(err => {
+              fileDiagnostics.push(this.createDiagnosticFromResult(err));
+            });
+            ForceSourceDeployExecutor._errorCollection.set(uri, fileDiagnostics);
+          });
+        } catch (e) {
+          // something else happened...?
+          console.log(`Could not parse Compile Errors`);
+        }
+      }
+    });
+
+    notificationService.reportExecutionError(
+      execution.command.toString(),
+      (execution.stderrSubject as any) as Observable<Error | undefined>
+    );
+    this.logMetric(execution.command.logName);
+    channelService.showChannelOutput();
+    channelService.streamCommandOutput(execution);
+    ProgressNotification.show(execution, cancellationTokenSource);
+    taskViewService.addCommandExecution(execution, cancellationTokenSource);
+  }
+
+  private getDeployResultData(stdErr: string) {
```

Comment:
will likely need to expand for different meta-data types

Reply:
CLI output will always be in the format `{ status: 0|1, result: /* command execution result */ }`. I think just doing a JSON.parse (with some error handling) might be enough.

Something I'm still looking into is if all possible errors are posted in the shape of the `Result` & `ForceSourceDeployErrorResult` interfaces you've added. So far they work for apex, custom objects and aura (with some info missing that might be a bug/gap). I'll get back to you once I'm able to confirm this.

Reply:
Luis is right here. There are some deploy errors that do not conform to the shape. We have to do a best case approximation – what happens if there is no file URI (or no line/column) to associate with those.

Reply:
Awesome, ya looks like it does work with objects deployments as well.

As mentioned below, I'm struggling to find a simple & consistent way to only grab the response JSON from stderr.  Here's an example stderr from a object deploy command.  You can see it has extra junk before the response. 

```
▸    sfdx-cli: update available from 6.27.0 to 6.31.0-32aa4b4408
[xmldom error]	end tag name: type maybe not complete 
@#[line:9,col:15]
{"message":"Deploy failed.","status":1,"stack":"DeployFailed: Deploy failed.\n    at ALMError (/Users/jonas/.local/share/sfdx/plugins/node_modules/salesforce-alm/dist/lib/almError.js:44:19)\n    at MetadataRegistry.initializeMetadataTypeInfos.then.catch.e (/Users/jonas/.local/share/sfdx/plugins/node_modules/salesforce-alm/dist/lib/source/sourceApiCommand.js:58:31)\n    at tryCatcher (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/util.js:16:23)\n    at Promise._settlePromiseFromHandler (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/promise.js:510:31)\n    at Promise._settlePromise (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/promise.js:567:18)\n    at Promise._settlePromise0 (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/promise.js:612:10)\n    at Promise._settlePromises (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/promise.js:687:18)\n    at Async._drainQueue (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/async.js:138:16)\n    at Async._drainQueues (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/async.js:148:10)\n    at Immediate.Async.drainQueues [as _onImmediate] (/Users/jonas/.local/share/sfdx/plugins/node_modules/bluebird/js/release/async.js:17:14)\n    at runCallback (timers.js:773:18)\n    at tryOnImmediate (timers.js:734:5)\n    at processImmediate [as _immediateCallback] (timers.js:711:5)","name":"DeployFailed","result":[{"columnNumber":"28","lineNumber":"13","error":"Error parsing file: 'asdasd' is not a valid value for the enum 'ActionOverrideType' (13:28)","fullName":"abc__c","type":"CustomObject","filePath":"force-app/main/default/objects/abc__c/abc__c.object-meta.xml"}],"warnings":[]}
```

Reply:
We usually use `--loglevel fatal` to silence any other information (see https://github.com/forcedotcom/salesforcedx-vscode/blob/develop/packages/salesforcedx-utils-vscode/src/cli/commandBuilder.ts#L63)

@amphro - Any thoughts on `[xmldom error]	end tag name: type maybe not complete `? I've not seen that one before.

Reply:
I'm using with `withJson()` which sets the `--loglevel fatal`, but I'm still getting the update and xml parsing junk... is this a bug with the CLI?

Reply:
I haven't seen `[xmldom error]	end tag name: type maybe not complete 
@#[line:9,col:15]` but the update available is well known and doesn't respect the `--loglevel fatal` either. We are picking up the story to always have the json response in stdout, which will simplify a lot of this. It should be fixed in the next month with an envvar to enable the new functionality to allow tools to slowly move over. 

Unfortunately, until that happens, there isn't a great workaround short of stipping everything before the first `{:` and hoping that is good enough.